pkg/operator/controller/controller_dns_daemonset.go

```diff
@@ -0,0 +1,190 @@
+	daemonset.Spec.Selector = DNSDaemonSetPodSelector(dns)
+	daemonset.Spec.Template.Labels = daemonset.Spec.Selector.MatchLabels
+
+	coreFileVolumeFound := false
+	for i := range daemonset.Spec.Template.Spec.Volumes {
+		// TODO: remove hardcoding of volume name
+		if daemonset.Spec.Template.Spec.Volumes[i].Name == "config-volume" {
+			daemonset.Spec.Template.Spec.Volumes[i].ConfigMap.Name = DNSConfigMapName(dns).Name
+			coreFileVolumeFound = true
+			break
+		}
+	}
+	if !coreFileVolumeFound {
+		return nil, fmt.Errorf("volume 'config-volume' is not found")
+	}
+
+	for i, c := range daemonset.Spec.Template.Spec.Containers {
+		switch c.Name {
+		case "dns":
+			daemonset.Spec.Template.Spec.Containers[i].Image = coreDNSImage
+		case "dns-node-resolver":
+			daemonset.Spec.Template.Spec.Containers[i].Image = openshiftCLIImage
+			envs := []corev1.EnvVar{}
+			if len(clusterIP) > 0 {
+				envs = append(envs, corev1.EnvVar{
+					Name:  "NAMESERVER",
+					Value: clusterIP,
+				})
+			}
+			if len(clusterDomain) > 0 {
+				envs = append(envs, corev1.EnvVar{
+					Name:  "CLUSTER_DOMAIN",
+					Value: clusterDomain,
+				})
+			}
+
+			if daemonset.Spec.Template.Spec.Containers[i].Env == nil {
+				daemonset.Spec.Template.Spec.Containers[i].Env = []corev1.EnvVar{}
+			}
+			daemonset.Spec.Template.Spec.Containers[i].Env = append(daemonset.Spec.Template.Spec.Containers[i].Env, envs...)
+		}
+	}
+	return daemonset, nil
+}
+
+// currentDNSDaemonSet returns the current dns daemonset.
+func (r *reconciler) currentDNSDaemonSet(dns *operatorv1.DNS) (*appsv1.DaemonSet, error) {
+	daemonset := &appsv1.DaemonSet{}
+	if err := r.client.Get(context.TODO(), DNSDaemonSetName(dns), daemonset); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return daemonset, nil
+}
+
+// createDNSDaemonSet creates a dns daemonset.
+func (r *reconciler) createDNSDaemonSet(daemonset *appsv1.DaemonSet) error {
+	if err := r.client.Create(context.TODO(), daemonset); err != nil {
+		return fmt.Errorf("failed to create dns daemonset %s/%s: %v", daemonset.Namespace, daemonset.Name, err)
+	}
+	logrus.Infof("created dns daemonset: %s/%s", daemonset.Namespace, daemonset.Name)
+	return nil
+}
+
+// updateDNSDaemonSet updates a dns daemonset.
+func (r *reconciler) updateDNSDaemonSet(current, desired *appsv1.DaemonSet) error {
+	changed, updated := daemonsetConfigChanged(current, desired)
+	if !changed {
+		return nil
+	}
+
+	if err := r.client.Update(context.TODO(), updated); err != nil {
+		return fmt.Errorf("failed to update dns daemonset %s/%s: %v", updated.Namespace, updated.Name, err)
+	}
+	logrus.Infof("updated dns daemonset: %s/%s", updated.Namespace, updated.Name)
+	return nil
+}
+
+// daemonsetConfigChanged checks if current config matches the expected config
+// for the cluster dns daemonset and if not returns the updated config.
+func daemonsetConfigChanged(current, expected *appsv1.DaemonSet) (bool, *appsv1.DaemonSet) {
```

Comment:
Note for the future: like with ingress, this function could be doing a more generic diff on `.spec` and if there's any diff simply overwrite the current `.spec.` with a copy of the expected `.spec`.